File: loader/src/main/java/im/mingxi/loader/hotpatch/HotPatch.java
package im.mingxi.loader.hotpatch;

import android.util.Log;
import dalvik.system.DexClassLoader;
import im.mingxi.loader.bridge.XPBridge;
import im.mingxi.loader.util.ActivityUtil;
import im.mingxi.loader.util.FileUtil;
import im.mingxi.loader.util.HttpUtil;
import im.mingxi.loader.util.PathUtil;
import java.io.File;
import java.io.IOException;
import java.lang.reflect.Method;
import java.nio.file.Files;
import java.nio.file.Path;

public class HotPatch {
    public static final String hotPatchPath = PathUtil.dataPath + "HotPatch/";
    public static final String hotPatchAPKPath = hotPatchPath + "release.apk";

    public static boolean onLoad() {
        String versions =
                HttpUtil.sendDataRequest("http://miao.yuexinya.top/HotPatch/versions.txt");
        if (versions != null && !"".equals(versions)) { // 正常连接到服务器
            if (getSign().equals(versions)) { // 签名一致，加载模块
                doLoad();
            } else { // 签名不一致，下载新版本
                try {
                    Path path = Path.of(hotPatchAPKPath);
                    File file = path.toFile();
                    if (!file.exists()) file.createNewFile();
                    HttpUtil.downloadToFile(
                            "http://miao.yuexinya.top/HotPatch/release.apk", hotPatchAPKPath);
                    /*懒得获取Context来调用 {@link im.mingxi.loader.util.ActivityUtil#killAppProcess(Context)}，所以我们暴力点*/
                    System.exit(0);
                } catch (Exception err) {
                    XPBridge.log(Log.getStackTraceString(err));
                }
            }
            return true;
        }
        return false;
    }

    private static void doLoad() {
        DexClassLoader dexClassLoader =
                new DexClassLoader(
                        hotPatchAPKPath,
                        hotPatchPath + "Optimized",
                        null,
                        HotPatch.class.getClassLoader());
        try {
            Class startupClass = dexClassLoader.loadClass("im.mingxi.miko.startup.StartUp");
            Method initMet = startupClass.getDeclaredMethod("doLoad");
            initMet.setAccessible(true);
            initMet.invoke(null);
        } catch (Exception err) {
            XPBridge.log(Log.getStackTraceString(err));
        }
    }


    private static String getSign() {
        return "(version = " + 666 + ")";
    }
}

File: loader/src/main/java/im/mingxi/loader/hotpatch/HotPatch.kt
package im.mingxi.loader.hotpatch


import android.util.Log
import dalvik.system.DexClassLoader
import im.mingxi.loader.bridge.XPBridge
import im.mingxi.loader.util.Constants
import im.mingxi.loader.util.FileUtil
import im.mingxi.loader.util.HttpUtil
import im.mingxi.loader.util.HttpUtil.sendDataRequest
import im.mingxi.loader.util.PathUtil
import java.io.File

object HotPatch {
    val hotPatchPath: String = PathUtil.dataPath + "HotPatch/"
    val hotPatchAPKPath: String = hotPatchPath + "release.apk"

    fun onLoad(): Boolean {
        if (!Constants.isHotPatch) {
            this.doLoadModuleLocal()
            return true
        }
        // 拉取云端版本
        val cloudVersionSign =
            sendDataRequest("\"http://miao.yuexinya.top/HotPatch/versions.txt\"") ?: return false
        if (this.getSign() == cloudVersionSign) { //签名正确则加载模块
            this.doLoadModuleCloud()
            return true
        }
        val file: File = File(hotPatchAPKPath)
        if (!file.exists()) file.createNewFile()
        HttpUtil.downloadToFile(
            "http://miao.yuexinya.top/HotPatch/release.apk", hotPatchAPKPath
        )
        val signFile = File("${hotPatchPath}sign.data")
        if (!signFile.exists()) signFile.createNewFile()
        signFile.writeText(cloudVersionSign)
        /*懒得获取Context来调用 {@link im.mingxi.loader.util.ActivityUtil#killAppProcess(Context)}，所以我们暴力点*/
        System.exit(0)


        return false
    }

    // Load Module from local
    // Only can invoke by debug mode
    private fun doLoadModuleLocal() {
        try {
            val startupClass: Class<*> = Class.forName("im.mingxi.miko.startup.StartUp")
            val initMet = startupClass.getDeclaredMethod("doLoad")
            true.also {
                initMet.isAccessible = it
            }
            initMet.invoke(null)
        } catch (err: Exception) {
            XPBridge.log(Log.getStackTraceString(err))
        }
    }

    // 拉取本地标签
    private fun getSign(): String {
        return FileUtil.readFileString("${hotPatchPath}sign.data") ?: "NO_SIGN"
    }

    // 加载从云端下载的dex并启动StartUp
    // 由StartUp自行完成res和so的加载
    private fun doLoadModuleCloud() {
        val dexClassLoader =
            DexClassLoader(
                hotPatchAPKPath,
                hotPatchPath + "Optimized",
                null,
                HotPatch.javaClass.classLoader
            )
        try {
            val startupClass = dexClassLoader.loadClass("im.mingxi.miko.startup.StartUp")
            val initMet = startupClass.getDeclaredMethod("doLoad")
            initMet.isAccessible = true
            initMet.invoke(null)
        } catch (err: java.lang.Exception) {
            XPBridge.log(Log.getStackTraceString(err))
        }
    }
}